tpm2/test/tpm2_test.go

```diff
@@ -2111,3 +2111,87 @@ func TestPolicyOr(t *testing.T) {
 		t.Errorf("PolicyOr with error: %v", err)
 	}
 }
+
+func TestNVUndefineSpaceSpecial(t *testing.T) {
+	rw := openTPM(t)
+	defer rw.Close()
+
+	// Set up policy session for creation in a trial session
+	sess, _, err := StartAuthSession(rw, HandleNull, HandleNull, make([]byte, 16), nil, SessionTrial, AlgNull, AlgSHA256)
+	if err != nil {
+		t.Errorf("StartAuthSession() for policy generation failed: %v", err)
+	}
+	// Use well known SHA256 values to generate a known policy.
+	hashVal := make([]byte, 32)
+	for iterator := range hashVal {
+		hashVal[iterator] = byte(iterator)
+	}
+	zeroDigest := make([]byte, 32)
+	// Use well known hash for policy generation with PolicyOr
+	err = PolicyOr(rw, sess, TPMLDigest{Digests: []tpmutil.U16Bytes{tpmutil.U16Bytes(hashVal), tpmutil.U16Bytes(zeroDigest)}})
+	if err != nil {
+		t.Errorf("PolicyOr() for policy generation failed: %v", err)
+	}
+
+	// Use PolicyCommandCode(rw, sess, CmdNVUndefineSpaceSpecial) to manipulate the session
+	err = PolicyCommandCode(rw, sess, CmdNVUndefineSpaceSpecial)
+	if err != nil {
+		t.Errorf("PolicyCommandCode() failed: %v", err)
+	}
+	// Read policy for NVDefineSpace
+	pol, err := PolicyGetDigest(rw, sess)
+	if err != nil {
+		t.Errorf("PolicyGetDigest() failed: %v", err)
+	}
+	// Close trail session
+	err = FlushContext(rw, sess)
+
+	// Test index with
+	pubIndex := NVPublic{
+		NVIndex:    0x1500000,
+		NameAlg:    AlgSHA256,
+		Attributes: AttrPlatformCreate | AttrPolicyDelete | AttrPolicyWrite | AttrNoDA | AttrAuthRead,
+		AuthPolicy: pol,
+		DataSize:   10,
+	}
+	// AuthCommand for platform hierarchy
+	platformAuthCmd := AuthCommand{
+		Session:    HandlePasswordSession,
+		Attributes: AttrContinueSession,
+		Auth:       EmptyAuth,
+	}
+
+	// Create the index
+	err = NVDefineSpaceEx(rw, HandlePlatform, "", pubIndex, platformAuthCmd)
+	if err != nil {
+		t.Errorf("NVDefineSpaceEx() failed: %v", err)
+	}
+
+	// Start new auth session for authentication of nv index access
+	sess, _, err = StartAuthSession(rw, HandleNull, HandleNull, make([]byte, 16), nil, SessionPolicy, AlgNull, AlgSHA256)
+	if err != nil {
+		t.Errorf("StartAuthSession() for nv index auth failed: %v", err)
+	}
+	// Use PolicyOr to manipulate the session with well known hash values
```

Comment:
So I am unable to reply to the earlier conversation here as it was on an old commit that got force-pushed over. I'll try to paraphrase what you said as now I can't even find the comment (oops!)

> The TPM resets the policy hash to the zeroPolicy, giving me control of the policy so I can use it for this test. See Part 3: Commands, section 23.6, PolicyOr: 'Reset policyDigest to a Zero Digest.' (and also Note 2)

To restate my main concern here: I don't think you need PolicyOR at all in order to test `UndefineSpaceSpecial`. But as TPM 2.0 policies in general (and PolicyOR in particular) are not explained super well in the spec, I'd like to also discuss this a bit as it may help.

Quick summary on policies (sorry if you don't need it 😊 ):

- Policies are expressed as simple hashes that can be extended, much like PCRs.
- Calling a `PolicySomething` command makes one or more assertions about the state of the TPM, and extends the current Policy hash
  - Example: `PolicyPCR` checks to see that the named PCR values are as expected, then extends some info about the PCRS that were checked into the current policy hash (and you can only get the TPM to extend this info by performing this check)
  - Example: `PolicyAuthValue` writes a bit for a future assertion that the command being authorized is also authorized by passing the object's auth value, and extends some info about the check into the current policy hash (and you can only get the TPM to extend this info by performing this check)
- Objects are authorized with policy by executing the exactly-correct policy in the exactly-correct order, and having all those commands succeed.

More details about `PolicyOR`:

- `PolicyOR` turns policies from a 1D array of ANDed-together assertions into a [Merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) of policies.
  - As with anything using a Merkle tree, in order to trace the path from the leaf (the particular policy you wanted to use this time) to the root (the actual policy hash on the object), you need some metadata about the rest of the tree. You don't need the whole rest of the tree, but you need all the peer nodes of all the nodes in your leaf-to-root traversal of the tree. i.e., you always have to pass the hashes of all the policies you COULD have satisfied, in order to satisfy one of them and end up with the final policy hash that matches the object whose policy is this tree you are trying to conform to.
- `PolicyOR` is an assertion-then-update-hash like any other `PolicySomething`. But in this case, the assertion is "the current policy hash is one of these up-to-8 values". And the extension logic is special (your note about the spec), because the whole point is to be able to get to the same hash value whether you used branch A or branch B of the `PolicyOR`.

So, the policy hash is _reset to the zeroPolicy_ and **then** extended with the information about **all** the policy branches (uniquely identifying the complete policy tree). That way, the same PolicyOR command executed on top of branchA or branchB results in the same hash - this wouldn't be possible if it simply extended the current policy hash value (or else you'd have to have a collision in your hash algorithm, which would make TPM policies useless garbage).

Hopefully that clarifies why PolicyOR does this reset. Now, to your PR 😁

This test is supposed to test `UndefineSpaceSpecial`. `UndefineSpaceSpecial` says:
> Because administrative role is required, the policy must contain a command that sets the policy command code to `TPM_CC_NV_UndefineSpaceSpecial`.

i.e., you are supposed to use `PolicyCommandCode` or `PolicyCpHash` to specifically announce "This policy is intended for use in authorizing `UndefineSpaceSpecial`! I'm OK with it!". The `Duplicate` command is similarly guarded against shooting yourself in the foot.

So, all you need to do to authorize this command for test is to have your test NV index have a simple policy that is just `PolicyCommandCode(TPM_CC_NV_UndefineSpaceSpecial)` (along with the needed attributes to disallow unspecial `UndefineSpace`.

The PolicyOr in this test is really just making the policy unnecessarily complex. It corresponds to this plain-language policy:

- You can authorize this command if:
  - Either
    - Nothing (the zero policy)
    - Something impossible (no known policy hashes to 0x000102030405...)
  - And
    - The command is UndefineSpaceSpecial (the PolicyCommandCode policy)

The whole "either nothing or something impossible" clause adds to the complexity of the test without adding test coverage (as this is not intended to be a PolicyOR test)

Reply:
Thank you good sir for that long explanation. I think i actually learned a lot about policies by it :)